src/Symfony/Bridge/Doctrine/Validator/DoctrineLoader.php

```diff
@@ -0,0 +1,121 @@
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bridge\Doctrine\Validator;
+
+use Doctrine\Common\Persistence\Mapping\MappingException;
+use Doctrine\ORM\EntityManagerInterface;
+use Doctrine\ORM\Mapping\ClassMetadataInfo;
+use Doctrine\ORM\Mapping\MappingException as OrmMappingException;
+use Symfony\Bridge\Doctrine\Validator\Constraints\UniqueEntity;
+use Symfony\Component\Validator\Constraints\Length;
+use Symfony\Component\Validator\Mapping\ClassMetadata;
+use Symfony\Component\Validator\Mapping\Loader\LoaderInterface;
+
+/**
+ * Guesses and loads the appropriate constraints using Doctrine's metadata.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+final class DoctrineLoader implements LoaderInterface
+{
+    private $entityManager;
+    private $classValidatorRegexp;
+
+    public function __construct(EntityManagerInterface $entityManager, string $classValidatorRegexp = null)
+    {
+        $this->entityManager = $entityManager;
+        $this->classValidatorRegexp = $classValidatorRegexp;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadClassMetadata(ClassMetadata $metadata): bool
+    {
+        $className = $metadata->getClassName();
+        if (null !== $this->classValidatorRegexp && !preg_match($this->classValidatorRegexp, $className)) {
+            return false;
+        }
+
+        try {
+            $doctrineMetadata = $this->entityManager->getClassMetadata($className);
+        } catch (MappingException | OrmMappingException $exception) {
+            return false;
+        }
+
+        if (!$doctrineMetadata instanceof ClassMetadataInfo) {
+            return false;
+        }
+
+        /* Available keys:
+           - type
+           - scale
+           - length
+           - unique
+           - nullable
+           - precision
+         */
+        $existingUniqueFields = $this->getExistingUniqueFields($metadata);
+
+        // Type and nullable aren't handled here, use the PropertyInfo Loader instead.
+        foreach ($doctrineMetadata->fieldMappings as $mapping) {
+            if (true === $mapping['unique'] && !isset($existingUniqueFields[$mapping['fieldName']])) {
+                $metadata->addConstraint(new UniqueEntity(array('fields' => $mapping['fieldName'])));
+            }
+
+            if (null === $mapping['length']) {
+                continue;
+            }
+
+            $constraint = $this->getLengthConstraint($metadata, $mapping['fieldName']);
+            if (null === $constraint) {
+                $metadata->addPropertyConstraint($mapping['fieldName'], new Length(array('max' => $mapping['length'])));
+            } elseif (null === $constraint->max) {
+                // If a Length constraint exists and no max length has been explicitly defined, set it
+                $constraint->max = $mapping['length'];
```

Comment:
I may be wrong but MySQL evaluates length with bytes instead of chars. Multi-bytes chars like with UTF-8 may use more than one byte and the Length constraint uses `mb_strlen` which counts chars instead of bytes.

We may need to set the `charset` option to `8bit`